rs/nervous_system/integration_tests/tests/upgrade_sns_controlled_canister_with_large_wasm.rs

```diff
@@ -0,0 +1,281 @@
+
+        assert!(uploaded_chunk_hashes.is_subset(&stored_chunk_hashes));
+        assert_eq!(uploaded_chunk_hashes.len(), num_chunks_expected);
+    }
+
+    uploaded_chunk_hashes
+}
+
+async fn run_test(store_same_as_target: bool) {
+    // 1. Prepare the world
+    let pocket_ic = PocketIcBuilder::new()
+        .with_nns_subnet()
+        .with_sns_subnet()
+        .with_application_subnet()
+        .build_async()
+        .await;
+
+    // Install the NNS canisters.
+    {
+        let with_mainnet_nns_canisters = false;
+        install_nns_canisters(&pocket_ic, vec![], with_mainnet_nns_canisters, None, vec![]).await;
+    }
+
+    // Publish SNS Wasms to SNS-W.
+    {
+        let with_mainnet_sns_canisters = false;
+        add_wasms_to_sns_wasm(&pocket_ic, with_mainnet_sns_canisters)
+            .await
+            .unwrap();
+    };
+
+    // Install a dapp canister.
+    let original_wasm = Wasm::from_bytes(very_large_wasm_bytes());
+    let original_wasm_hash = original_wasm.sha256_hash();
+
+    let app_subnet = pocket_ic.topology().await.get_app_subnets()[0];
+
+    let target_canister_id = install_canister_on_subnet(
+        &pocket_ic,
+        app_subnet,
+        vec![],
+        Some(original_wasm.clone()),
+        vec![ROOT_CANISTER_ID.into()],
+    )
+    .await;
+
+    let sns = {
+        let create_service_nervous_system = CreateServiceNervousSystemBuilder::default()
+            .with_dapp_canisters(vec![target_canister_id])
+            .build();
+
+        let swap_parameters = create_service_nervous_system
+            .swap_parameters
+            .clone()
+            .unwrap();
+
+        let sns_instance_label = "1";
+        let (sns, _) = nns::governance::propose_to_deploy_sns_and_wait(
+            &pocket_ic,
+            create_service_nervous_system,
+            sns_instance_label,
+        )
+        .await;
+
+        sns::swap::await_swap_lifecycle(&pocket_ic, sns.swap.canister_id, Lifecycle::Open)
+            .await
+            .unwrap();
+        sns::swap::smoke_test_participate_and_finalize(
+            &pocket_ic,
+            sns.swap.canister_id,
+            swap_parameters,
+        )
+        .await;
+
+        sns
+    };
+
+    let store_canister_id = if store_same_as_target {
+        target_canister_id
+    } else {
+        install_canister_on_subnet(
+            &pocket_ic,
+            app_subnet,
+            vec![],
+            None,
+            vec![sns.root.canister_id],
+        )
+        .await
+    };
+
+    let new_wasm = {
+        let new_wasm_bytes = modify_wasm_bytes(&original_wasm.bytes(), 42);
+        Wasm::from_bytes(&new_wasm_bytes[..])
+    };
+    let new_wasm_hash = new_wasm.sha256_hash();
+
+    // Smoke test
+    assert_ne!(new_wasm_hash, original_wasm_hash);
+
+    // WASM with 15_843_866 bytes (`image-classification.wasm.gz`) is split into 1 MiB chunks.
+    let num_chunks_expected = 16;
```

Comment:
This is going to break when that wasm gets bigger than 16_000_000 bytes.  Maybe do some math here, but also assert it's greater than 2MB (i.e. needs chunks)?

Reply:
This particular wasm is checked into the repo, so it's unlikely that it will change often. It seems to only be there for testing, previously, in the EE crates, and now here.

Reply:
Let me close this and merge as-is, I promise to follow the suggestion if this assertion fires at least once.